src/vp8.rs

```diff
@@ -1684,14 +1692,15 @@ impl<R: Read> Vp8Decoder<R> {
                 literal @ DCT_1..=DCT_4 => i16::from(literal),
 
                 category @ DCT_CAT1..=DCT_CAT6 => {
-                    let t = PROB_DCT_CAT[(category - DCT_CAT1) as usize];
+                    let probs = PROB_DCT_CAT[(category - DCT_CAT1) as usize];
 
                     let mut extra = 0i16;
-                    let mut j = 0;
 
-                    while t[j] > 0 {
-                        extra = extra + extra + self.partitions[p].read_bool(t[j])? as i16;
-                        j += 1;
+                    for t in probs.iter().copied() {
+                        if t == 0 {
+                            break;
+                        }
+                        extra = extra + extra + reader.read_bool(t)? as i16;
```

Comment:
One vague idea here: This is a tight loop and yet control flow by `?` probably turns it into a rather subtle analysis nightmare for llvm. At the same time there's a small, fixed maximum iteration count (11) and in many specialized cases of `category` that value is even lower. Is it possible to _buffer_ suitably many bits in the reader in such a way that there's only *one* exhaustion check and the rest of bits are just read from memory?

If the check is after the loop, it doesn't even matter if we error by double-bookkeeping, i.e. comparing number of available with consumed bits. It only matters that _when the read succeeds_ all bits must be accurate. In the error case, previously read bools can be arbitrarily chosen even to values that didn't actually appear in the input stream (due to it being too short).

Reply:
Another vague idea: `DecodingError` is 32 bytes (256 bits) in size according to `mem::size_of`, so it's probably not great to return it from a hot function. It would have to get spilled onto the stack instead of getting returned in a register. But I don't know how that interacts with inlining, maybe it doesn't even matter when the function is inlined.

Reply:
The `read_bool` method is [more complicated than you might guess](https://github.com/image-rs/image-webp/blob/5d8654dd160549777729a3888cb28172d450e078/src/vp8.rs#L702). I don't expect it would be possible to buffer future bits or to easily rewind it when those bits aren't necessary.

The size of `DecodingError` could be worth looking into. In this function, I think the only possible error is `io::Error` so we could cut the size in half by switching the signatures to use that instead. It would be a breaking change, but we may also want to cut down the size of `DecodingError` by removing the two variants that hold `String` types and replacing them either with either `&'static str` or just dedicated variants for each case

Reply:
I tried trimming the type down to 16 bytes by boxing the `String`s and it didn't make a difference on end-to-end benchmarks.